vault/identity_store_oidc_provider.go

```diff
@@ -2545,51 +2551,83 @@ func (i *IdentityStore) storeOIDCDefaultResources(ctx context.Context, view logi
 		i.Logger().Debug("wrote OIDC default provider")
 	}
 
-	// Store the default key
-	storageKey = namedKeyConfigPath + defaultKeyName
+	if _, err := i.ensureDefaultKey(ctx, view); err != nil {
+		return fmt.Errorf("error writing default key to storage: %w", err)
+	}
+
+	// Store the allow all assignment
+	storageKey = assignmentPath + allowAllAssignmentName
 	entry, err = view.Get(ctx, storageKey)
 	if err != nil {
 		return err
 	}
 	if entry == nil {
-		defaultKey := defaultOIDCKey()
-
-		// Generate initial key material for current and next keys
-		err = defaultKey.generateAndSetKey(ctx, i.Logger(), view)
-		if err != nil {
-			return err
-		}
-		err = defaultKey.generateAndSetNextKey(ctx, i.Logger(), view)
-		if err != nil {
-			return err
-		}
-
-		// Store the entry
-		entry, err := logical.StorageEntryJSON(storageKey, defaultKey)
+		entry, err := logical.StorageEntryJSON(storageKey, allowAllAssignment())
 		if err != nil {
 			return err
 		}
 		if err := view.Put(ctx, entry); err != nil {
 			return err
 		}
-		i.Logger().Debug("wrote OIDC default key")
+		i.Logger().Debug("wrote OIDC allow_all assignment")
 	}
 
-	// Store the allow all assignment
-	storageKey = assignmentPath + allowAllAssignmentName
-	entry, err = view.Get(ctx, storageKey)
+	return nil
+}
+
+func (i *IdentityStore) ensureDefaultKey(ctx context.Context, storage logical.Storage) (*namedKey, error) {
+	key, err := i.getNamedKey(ctx, storage, defaultKeyName)
+	if err != nil {
+		return nil, err
+	}
+	if key != nil {
+		return key, nil
+	}
+
+	// The default key doesn't exist. Write it to storage.
+	defaultKey := defaultOIDCKey()
+	entry, err := logical.StorageEntryJSON(namedKeyConfigPath+defaultKeyName, defaultKey)
+	if err != nil {
+		return nil, err
+	}
+	if err := storage.Put(ctx, entry); err != nil {
+		return nil, err
+	}
+
+	i.Logger().Debug("wrote OIDC default key")
+	return &defaultKey, nil
+}
+
+func (i *IdentityStore) lazyGenerateDefaultKey(ctx context.Context, storage logical.Storage) error {
```

Comment:
Not sure if I'm missing it elsewhere, but do we need some locking to ensure we don't call this twice in a race? Unlikely to happen with a human operator, but could possibly happen for scripted scenarios or during Terraform provider configuration?

Reply:
Yep! I found a data race when calling this concurrently in isolation. I think we're safe though because we hold write locks where this is called in our API handlers at [identity_store_oidc.go#L1153-L1154](https://github.com/hashicorp/vault/blob/main/vault/identity_store_oidc.go#L1153-L1154) and [identity_store_oidc_provider.go#L1067-L1068](https://github.com/hashicorp/vault/blob/main/vault/identity_store_oidc_provider.go#L1067-L1068).

I've added a comment to call out that the write lock must be held in https://github.com/hashicorp/vault/pull/24332/commits/9a1d69f716c1a4deb07b37cce4d773b376e6bda3.

Reply:
Awesome! Sorry for the false positive 👍